org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/NewProjectWizardTemplateParameterPage.java

```diff
@@ -0,0 +1,165 @@
+				variableWidgets.add(Tuples.create(stringVariable, text));
+				text.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));
+				text.setText(stringVariable.getValue());
+				text.setToolTipText(variable.getDescription());
+				text.addModifyListener(e -> {
+					stringVariable.setValue(text.getText());
+					update();
+				});
+			} else if (variable instanceof StringSelectionProjectVariable) {
+				StringSelectionProjectVariable selectionVariable = (StringSelectionProjectVariable) variable;
+				Combo combo = new Combo(main, SWT.READ_ONLY);
+				variableWidgets.add(Tuples.create(selectionVariable, combo));
+				combo.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));
+				combo.setItems(selectionVariable.getPossibleValues());
+				combo.setText(selectionVariable.getValue());
+				combo.setToolTipText(variable.getDescription());
+				combo.addSelectionListener(new SelectionAdapter() {
+					@Override
+					public void widgetSelected(SelectionEvent e) {
+						selectionVariable.setValue(combo.getText());
+						update();
+					}
+				});
+			} else if (variable instanceof BooleanProjectVariable) {
+				BooleanProjectVariable booleanVariable = (BooleanProjectVariable) variable;
+				Button checkbox = new Button(main, SWT.CHECK);
+				variableWidgets.add(Tuples.create(booleanVariable, checkbox));
+				checkbox.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false));
+				checkbox.setSelection(booleanVariable.getValue());
+				checkbox.setToolTipText(variable.getDescription());
+				checkbox.addSelectionListener(new SelectionAdapter() {
+					@Override
+					public void widgetSelected(SelectionEvent e) {
+						booleanVariable.setValue(checkbox.getSelection());
+						update();
+					}
+				});
+			}
+		}
+		update();
+		setControl(main);
+	}
+
+	private void update() {
+		if (inUserAction) {
+			inUserAction = false;
+			refreshVariables();
+			validate();
+			inUserAction = true;
+		}
+	}
+
+	protected void validate() {
+		IStatus status = projectTemplate.validate();
+		if (status == null || status.getSeverity() == IStatus.OK) {
+			setErrorMessage(null);
+			setMessage(null);
+			setPageComplete(true);
+		} else if (status.getSeverity() == IStatus.ERROR) {
+			setErrorMessage(status.getMessage());
+			setPageComplete(false);
+		} else if (status.getSeverity() == IStatus.WARNING) {
+			setErrorMessage(null);
+			setMessage(status.getMessage(), IMessageProvider.WARNING);
+			setPageComplete(true);
+		} else {
+			setErrorMessage(null);
+			setMessage(status.getMessage(), IMessageProvider.INFORMATION);
+			setPageComplete(true);
+		}
+	}
+
+	private void refreshVariables() {
+		projectTemplate.updateVariables();
+		for (Pair<ProjectVariable, Control> pair : variableWidgets) {
+			ProjectVariable variable = pair.getFirst();
+			Widget widget = pair.getSecond();
+			if (variable instanceof StringProjectVariable) {
+				StringProjectVariable stringVariable = (StringProjectVariable) variable;
```

Comment:
Can this logic be moved into the ProjectVariable? This would allow clients to specialize the logic by providing subtypes of the e.g. StringProjectVariable

Reply:
Will move that (see above).